cairo/crates/contracts/src/hooks/interchain_gas_paymaster.cairo

```diff
@@ -0,0 +1,404 @@
+                        Errors::INVALID_METADATA,
+                    );
+                    StandardHookMetadata::gas_limit(_metadata, DEFAULT_GAS_USAGE)
+                },
+            };
+            let destination_domain = _message.destination;
+            self.quote_gas_payment(destination_domain, gas_limit)
+        }
+    }
+
+    // IInterchainGasPaymaster IMPLEMENTATION
+    #[abi(embed_v0)]
+    impl IInterchainGasPaymasterImpl of IInterchainGasPaymaster<ContractState> {
+        /// Pays `payment` native tokens (ERC20 ETH token) for gas.
+        ///
+        /// # Arguments
+        ///
+        /// * `_message_id` - ID of the message to pay for
+        /// * `_destination_domain` - Domain ID of the destination chain
+        /// * `_gas_amount` - Amount of gas to pay for on the destination chain
+        /// * `_payment` - Amount of native tokens to pay
+        /// * `_refund_address` - Address to refund any overpayment
+        fn pay_for_gas(
+            ref self: ContractState,
+            _message_id: u256,
+            _destination_domain: u32,
+            _gas_amount: u256,
+            _payment: u256,
+            _refund_address: ContractAddress,
+        ) {
+            assert(_refund_address != contract_address_const::<0>(), Errors::ZERO_REFUND_ADDRESS);
+
+            let required = self.quote_gas_payment(_destination_domain, _gas_amount);
+            assert(_payment >= required, Errors::INSUFFICIENT_PAYMENT);
+
+            let caller = get_caller_address();
+            let gas_token = ERC20ABIDispatcher { contract_address: ETH_ADDRESS() };
+            gas_token.transfer_from(caller, get_contract_address(), _payment);
+
+            let refund_amount = _payment - required;
+            if refund_amount > 0 {
+                gas_token.transfer( _refund_address, refund_amount);
+                self.emit(GasRefund { refund_address: _refund_address, amount: refund_amount });
+            }
+
+            self.emit(GasPayment { message_id: _message_id, destination_domain: _destination_domain, gas_amount: _gas_amount, payment: required });
+        }
+
+        /// Quotes gas payment amount for a given destination domain and gas amount.
+        ///
+        /// # Arguments
+        ///
+        /// * `_destination_domain` - Domain ID of the destination chain
+        /// * `_gas_amount` - Amount of gas to quote for on the destination chain
+        ///
+        /// # Returns
+        ///
+        /// u256 - Total required payment in native tokens
+        fn quote_gas_payment(
+            ref self: ContractState,
+            _destination_domain: u32,
+            _gas_amount: u256,
+        ) -> u256 {
+            let config = self.gas_configs.read(_destination_domain);
+            assert(config.token_exchange_rate != 0, Errors::CONFIG_NOT_FOUND);
+            let total_gas: u256 = _gas_amount + config.gas_overhead;
```

Comment:
users calling quote_gas_payment are expecting a quote for a specific gas limit, we shouldn't be adding the gas overhead.